pkg/upgrade/upgrade_state.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright 2022 NVIDIA CORPORATION & AFFILIATES
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Mellanox/network-operator/api/v1alpha1"
+	"github.com/Mellanox/network-operator/pkg/consts"
+)
+
+type NodeUpgradeState struct {
+	Node            *v1.Node
+	DriverPod       *v1.Pod
+	DriverDaemonSet *appsv1.DaemonSet
+}
+
+type ClusterUpgradeState struct {
+	Policy *v1alpha1.OfedUpgradePolicySpec
+
+	NodeStates map[string][]*NodeUpgradeState
+}
+
+func NewClusterUpgradeState() ClusterUpgradeState {
+	return ClusterUpgradeState{NodeStates: make(map[string][]*NodeUpgradeState)}
+}
+
+type ClusterUpgradeStateManager struct {
+	K8sClient         client.Client
+	Log               logr.Logger
+	DrainManager      DrainManager
+	PodRestartManager PodRestartManager
+}
+
+func NewClusterUpdateStateManager(
+	drainManager DrainManager,
+	podRestartManager PodRestartManager,
+	log logr.Logger,
+	k8sClient client.Client) *ClusterUpgradeStateManager {
+	manager := &ClusterUpgradeStateManager{
+		DrainManager:      drainManager,
+		PodRestartManager: podRestartManager,
+		Log:               log,
+		K8sClient:         k8sClient,
+	}
+
+	return manager
+}
+
+func (m *ClusterUpgradeStateManager) ApplyState(ctx context.Context, currentState *ClusterUpgradeState) {
+	m.Log.V(consts.LogLevelInfo).Info("State Manager, got state update")
+
+	policy := currentState.Policy
+	if !policy.AutoUpgrade {
+		m.Log.V(consts.LogLevelInfo).Info("Driver auto upgrade is disabled, skipping")
+		return
+	}
+
+	m.Log.V(consts.LogLevelInfo).Info("Node states:",
```

Comment:
I think  we should add state validation logic at the beginning of this function. This logic should validate that annotation which was read from the Node object is really reflect current status. In this validation handler it should be possible to directly move node between states and update node annotation if we detect that information is outdated. Node annotation can contain outdated information for example if operator restarts or if we failed to update node annotation on the previous cycle.

Reply:
Was discussed offline. We agreed that each stage can optionally include extra steps to validate current state before applying any actions. No need for generic validation function.